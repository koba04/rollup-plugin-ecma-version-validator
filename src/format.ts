export const format = (
  error: { message: string; loc?: { line: number; column: number } },
  source: string
): string => {
  const loc = error.loc;
  let message = error.message;
  if (loc) {
    message += `\n[${loc.line}:${loc.column}] ${extractSource(source, loc)}`;
  }
  return message;
};

const extractSource = (
  source: string,
  loc: { line: number; column: number }
) => {
  const offset = 40;
  const sourceLine = source.split("\n")[loc.line - 1];
  if (sourceLine.length < 100) {
    return sourceLine;
  }
  return sourceLine.substring(loc.column - offset, loc.column + offset);
};
