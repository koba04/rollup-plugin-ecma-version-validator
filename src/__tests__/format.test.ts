import { format } from "../format";

describe("format", () => {
  it("should return a formatted message", () => {
    const source = "var a = true;\nconst foo = () => {}";
    const error = {
      message: "const is a reserved word",
      loc: {
        line: 2,
        column: 0,
      },
    };
    expect(format(error, source)).toMatchSnapshot();
  });
  it("should return a formatted message if loc is not defined", () => {
    const source = "var a = true;\nconst foo = () => {}";
    const error = {
      message: "const is a reserved word",
    };
    expect(format(error, source)).toMatchSnapshot();
  });
  it("should omit a source code if the error line is over 100 characters", () => {
    const source = `var ${"a".repeat(100)} = () => {}; // ${"b".repeat(200)}`;
    const error = {
      message: "const is a reserved word",
      loc: {
        line: 1,
        column: 110,
      },
    };
    expect(format(error, source)).toMatchSnapshot();
  });
});
