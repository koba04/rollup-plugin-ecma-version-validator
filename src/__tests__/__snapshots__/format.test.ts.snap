// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format should omit a source code if the error line is over 100 characters 1`] = `
"const is a reserved word
[1:110] aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = () => {}; // bbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
`;

exports[`format should return a formatted message 1`] = `
"const is a reserved word
[2:0] const foo = () => {}"
`;

exports[`format should return a formatted message if loc is not defined 1`] = `"const is a reserved word"`;
